### PR TITLE
Flink: Avoid RANGE mode broken chain when write parallelism changes

### PR DESCRIPTION
**Pull request by huyuanfeng2018**
Detailed： #11543

This fix will modify the data processing process in RANGE mode, set the filter link of `StatisticsOrRecord` and actively set it to `writeParallelism` to avoid the shuffle mode changing to rebalance due to changes in parallelism in subsequent write operators.

### CODE REVIEW

**Comment by huyuanfeng2018** [comment]
@pvary  @stevenzwu Can you help me review it if you have time? Thanks a lot

**Comment by pvary** [comment]
@huyuanfeng2018: Could we have a unit test for this to avoid future regressions?

**Comment by huyuanfeng2018** [comment]
> @huyuanfeng2018: Could we have a unit test for this to avoid future regressions?

Thanks  @pvary and @mxm  review~

`TestFlinkIcebergSinkDistributionMode` add unit test under different writeParallelism .

@stevenzwu cc, Maybe you will be interested in this~

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java [hunk 1]
```diff
@@ -666,8 +667,16 @@ private DataStream<RowData> distributeDataStream(
...
-              .filter(StatisticsOrRecord::hasRecord)
-              .map(StatisticsOrRecord::record);
+              .flatMap(
+                  (FlatMapFunction<StatisticsOrRecord, RowData>)
+                      (statisticsOrRecord, out) -> {
+                        if (statisticsOrRecord.hasRecord()) {
+                          out.collect(statisticsOrRecord.record());
+                        }
+                      })
+              // Set the parallelism same as writerParallelism avoid broken chain
```
Comment (stevenzwu) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:677]: nit: maybe this?
```
Set the parallelism same as writerParallelism to promote operator chaining with the downstream writer operator.
```

**Review by stevenzwu — APPROVED** [review]
just a nit comment

**Comment by huyuanfeng2018** [comment]
> just a nit comment

fixed. PTAL

**Comment by stevenzwu** [comment]
thanks @huyuanfeng2018

**Comment by stevenzwu** [comment]
@huyuanfeng2018 can you create a back port for this fix?

**Comment by huyuanfeng2018** [comment]
Thanks @Stevenzwu Ping, I will open a PR later. ~

**Pull request merged**